File: WorkingInProgress/tuttoConPygame/crazyGoose.py
from Giocatore_modul import * 
from Percorso_modul import *
from Casella_modul import *

import tkinter as tk
import pygame
import random
import time
import threading

INFO_DADO_COM = "Dado COM: "
	
class Dado():
	def __init__(self):
		random.seed()

	def tiraDado(self):
			#Tira il dado, ossia un numero casuale tra 1 e 6 (estremi INCLUSI)
		self.dado = random.randint(1,6)
		return self.dado

class Button:
	def __init__(self, crazyGoose, game, x, y, width, height, text):
		self.crazyGoose = crazyGoose
		self.game = game
		self.x = x
		self.y = y
		self.width = width
		self.height = height
		self.rect = pygame.Rect(x, y, width, height)
		
		pygame.draw.rect(self.game.display, self.game.BLACK, self.rect, 1, 4)
		self.write_text(text, width, height, x, y)
	
	def write_text(self, text, width, height, x, y):
		font = pygame.font.Font(self.game.font_name, 13)
		text_surface = font.render(text, True, self.game.BLACK)
		text_rect = text_surface.get_rect()
		text_rect.center = ((x+width/2), (y+height/2))
		self.game.display.blit(text_surface, text_rect)
	
	def pressed(self, mouse):
		mouse_x = mouse[0]
		mouse_y = mouse[1]
		
		if(self.x <= mouse_x <= self.x+self.width
				and self.y <= mouse_y <= self.y + self.height):
			self.crazyGoose.tiraDado()
	
	def detectMouseOver(self, mouse):
		if(self.rect.collidepoint(mouse)):
			#codice per "evidenziare" il rettangolo (per far capire all'utente che ci sta passando sopra)
			pass
	
class CrazyGoose():
	def __init__(self, game):
		self.game = game
		self.gameStarted = False
		self.buttonDadoPL1 = None
		
	# sovrappone
	def blit_screen(self):
		self.game.window.blit(self.game.display, (0, 0))
		pygame.display.update()
	
	def start(self):
		if(self.gameStarted == False):
			self.gameStarted = True
			
			self.partitaTerminata = False
	
			self.creaWidgets()
			self.blit_screen()
	
	def mousePressed(self, mousePosition):
		if(self.buttonDadoPL1 != None):
			self.buttonDadoPL1.pressed(mousePosition)
	
	def mouseOver(self, mousePosition):
		self.buttonDadoPL1.detectMouseOver(mousePosition)
	
	def creaWidgets(self):
			#"ripulisce" tutto
		self.game.display.fill(self.game.WHITE)
			#lista di oggetti Casella
		self.caselle = list()
		self.percorso = Percorso()
		
			#"DISEGNA" gli ellissi
		self.posizionaLeCaselle()
			#scrive nelle caselle il loro effetto
		self.riempiCaselle()
		
		self.draw_text("Tu (PL1)", 15, 50, 20)
		self.draw_text("Dado PL1", 14, 55, 60)
		
		self.buttonDadoPL1 = Button(self, self.game, 100, 43, 35, 35, "0")
		
		self.draw_text("Computer (COM)", 14, 930, 20)
		self.draw_text((INFO_DADO_COM+"0"), 14, 917, 60)
		
		self.player = Giocatore(self.game, self.caselle, self.percorso, "PL1")
		self.com = Giocatore(self.game, self.caselle, self.percorso, "COM")
		
		
		#Decide chi incomincia, tira il dado e vede se il numero tirato è pari o dispari
			# (tra 1 e 6 ci sono 3 pari e 3 dispari, perciò 50% possbilità a testa)
		if(Dado().tiraDado() % 2 == 0):
			self.player.turnoMio = True
			self.com.turnoMio = False
			
				#PER ORA cambia il background-color della label in alto
			self.segnalaChiTocca(True)
		else:
			self.player.turnoMio = False
			self.com.turnoMio = True
			
			turnoCOM = threading.Thread(target=self.toccaAlCOM)
			turnoCOM.start()
				
				#PER ORA cambia il background-color della label in alto
			self.segnalaChiTocca(False)

	def tiraDado(self):
			#Se la partita è terminata "blocca" il dado
		if(not self.partitaTerminata):
			numEstratto = Dado().tiraDado()
			if(self.player.turnoMio):
				self.avanzaPlayer1(numEstratto)
			else:
				turnoCOM = threading.Thread(target=self.toccaAlCOM, args=(numEstratto,))
				turnoCOM.start()
	

	def toccaAlCOM(self, numEstratto=-1):
		if(numEstratto == -1):
			numEstratto = Dado().tiraDado()
		
			#Se il COM ha un fermo entrerà in avanzaPlayer2, abbasserà il suo fermo,
			# e eseguirà avanzaPlayer1, perciò non voglio che il gioco si fermi per nulla
		if(self.com.turniFermo == 0):
			time.sleep(2)

		self.avanzaPlayer2(numEstratto)


	def avanzaPlayer1(self, numEstratto):
			#Se il PL1 ha beccato un fermo in precedenza deve "consumarlo"
			# (sarà come se avesse tirato l'avversario)
		if(self.player.turniFermo > 0):
			self.player.turniFermo -= 1
			
			self.player.turnoMio = False
			self.com.turnoMio =	 True

			self.segnalaChiTocca(False)

			turnoCOM = threading.Thread(target=self.toccaAlCOM, args=(numEstratto,))
			turnoCOM.start()
		else:
			"""
			TODO
			self.valDadoPL1.set(numEstratto)
			"""

			toccaAncoraA_Me = self.player.avanza(numEstratto)
			if(not self.player.vincitore):
					#avanza() ritorna self.turnoMio, perciò
					# se True non tocca all'avversario (gli setto False)
					# se False tocca all'avversario (gli setto True)
				self.com.turnoMio = not toccaAncoraA_Me
					
					#se prende un fermo ANNULLA il fermo dell'avversario 
					#(SENNÒ NESSUNO GIOCHEREBBE PIÙ per tot turni)
				if(self.player.turniFermo > 0):
					self.com.turniFermo = 0
						#Ora il PL1 ha un fermo, quindi tocca all'avversario
					self.segnalaChiTocca(False)
					self.tiraDado()
				else:
					if(self.com.turniFermo > 0):
							#Se l'avversario ha un fermo al 100% tocca al PL1...
						self.segnalaChiTocca(True)
					else:
						#L'avversario non ha un fermo MA non è detto che tocchi a lui 
						# (PL1 potrebbe aver preso un TIRA DI NUOVO)
						if(self.com.turnoMio):
							self.segnalaChiTocca(False)

							turnoCOM = threading.Thread(target=self.toccaAlCOM)
							turnoCOM.start()
						else:
							self.segnalaChiTocca(True)

			else:
					#Segnala PL1 come vincitore e fa TERMINARE la partita
				self.segnalaVincitore(True)
				self.partitaTerminata = True


	def avanzaPlayer2(self, numEstratto):
			#Se il COM ha beccato un fermo in precedenza deve "consumarlo"
			# (sarà come se avesse tirato l'avversario)
		if(self.com.turniFermo > 0):
			self.com.turniFermo -= 1

			self.player.turnoMio = True
			self.com.turnoMio = False

			self.segnalaChiTocca(True)

			self.avanzaPlayer1(numEstratto)
		else:
			"""
			TODO
			self.valDadoCOM.set( (INFO_DADO_COM+str(numEstratto)) )
			"""

			toccaAncoraA_Me = self.com.avanza(numEstratto)
			
			if(not self.com.vincitore):
					#avanza() ritorna self.turnoMio, perciò
					# se True non tocca all'avversario (gli setto False)
					# se False tocca all'avversario (gli setto True)
				self.player.turnoMio = not toccaAncoraA_Me
				
					#se prende un fermo ANNULLA il fermo dell'avversario 
					#(SENNÒ NESSUNO GIOCHEREBBE PIÙ per tot turni)
				if(self.com.turniFermo > 0):
						#Ora il COM ha un fermo, quindi tocca all'avversario
					self.player.turniFermo = 0
					self.segnalaChiTocca(True)
				else:
					if(self.player.turniFermo > 0):
							#Se l'avversario ha un fermo al 100% tocca al COM, lo segnala
						self.segnalaChiTocca(False)
						self.tiraDado()
					else:
						#L'avversario non ha un fermo MA non è detto che tocchi a lui 
						# (COM potrebbe aver preso un TIRA DI NUOVO)
						if(self.player.turnoMio):
							self.segnalaChiTocca(True)
						else:
							self.segnalaChiTocca(False)
							turnoCOM = threading.Thread(target=self.toccaAlCOM)
							turnoCOM.start()
			else:
					#Segnala COM come vincitore e fa TERMINARE la partita
				self.segnalaVincitore(False)
				self.partitaTerminata = True


	def segnalaChiTocca(self, turnoPlayer1):
		pass
		"""if(turnoPlayer1):
			self.lblInfoPl.configure(bg="red")
			self.lblInfoCom.configure(bg="grey")
		else:
			self.lblInfoPl.configure(bg="grey")
			self.lblInfoCom.configure(bg="red")"""
	
	def segnalaVincitore(self, haVintoPlayer1):
		pass
		"""if(haVintoPlayer1):
			self.lblInfoPl.configure(bg="orange")
			self.lblInfoCom.configure(bg="grey")
		else:
			self.lblInfoPl.configure(bg="grey")
			self.lblInfoCom.configure(bg="orange")"""

			
	def riempiCaselle(self):
		"""Cicla per ogni casella e controlla, se la posizione della casella
		 c'è tra le chiavi del dizionario del percorso allora "carico" l'effetto
		 (prendendo il codice contenuto nel valore con quella chiave, il codice della casella),
		 altrimenti non metto nulla (uso il codice -1)"""
		i=1
		for i in range(QTA_CASELLE_TOTALI+1):
			if(i in self.percorso.dictCaselle.keys()):
				self.caselle[i-1].settaEffetto(self.percorso.dictCaselle[i], self.game.display, self.game.font_name)
			else:	#chiave inesistente (cioè posizione VUOTA)
				self.caselle[i-1].settaEffetto(-1, self.game.display, self.game.font_name)


	def posizionaLeCaselle(self):
		self.caselle.append( Casella(self.game.display, 160, 600) )
		self.caselle.append( Casella(self.game.display, 270, 600) )
		self.caselle.append( Casella(self.game.display, 380, 600) )
		self.caselle.append( Casella(self.game.display, 490, 600) )
		self.caselle.append( Casella(self.game.display, 600, 600) )
		self.caselle.append( Casella(self.game.display, 710, 600) )
		
		self.caselle.append( Casella(self.game.display, 790, 545) )
		self.caselle.append( Casella(self.game.display, 840, 475) )
		self.caselle.append( Casella(self.game.display, 850, 405) )
		self.caselle.append( Casella(self.game.display, 840, 335) )
		self.caselle.append( Casella(self.game.display, 790, 265) )
		
		self.caselle.append( Casella(self.game.display, 710, 210) )
		self.caselle.append( Casella(self.game.display, 600, 180) )
		self.caselle.append( Casella(self.game.display, 490, 180) )
		self.caselle.append( Casella(self.game.display, 380, 180) )
		self.caselle.append( Casella(self.game.display, 270, 180) )
		self.caselle.append( Casella(self.game.display, 170, 220) )

		self.caselle.append( Casella(self.game.display, 125, 300) )
		self.caselle.append( Casella(self.game.display, 125, 385) )
		self.caselle.append( Casella(self.game.display, 125, 470) )

		self.caselle.append( Casella(self.game.display, 215, 530) )
		self.caselle.append( Casella(self.game.display, 325, 530) )
		self.caselle.append( Casella(self.game.display, 435, 530) )
		self.caselle.append( Casella(self.game.display, 545, 530) )
		self.caselle.append( Casella(self.game.display, 653, 515) )
		
		self.caselle.append( Casella(self.game.display, 735, 465) )
		self.caselle.append( Casella(self.game.display, 735, 380) )
		self.caselle.append( Casella(self.game.display, 680, 305) )
		
		self.caselle.append( Casella(self.game.display, 585, 260) )
		self.caselle.append( Casella(self.game.display, 475, 260) )
		self.caselle.append( Casella(self.game.display, 365, 260) )
		
		self.caselle.append( Casella(self.game.display, 265, 295) )
		self.caselle.append( Casella(self.game.display, 230, 370) )
		self.caselle.append( Casella(self.game.display, 270, 445) )
		
		self.caselle.append( Casella(self.game.display, 380, 460) )
		self.caselle.append( Casella(self.game.display, 490, 460) )
		self.caselle.append( Casella(self.game.display, 590, 440) )
		
		self.caselle.append( Casella(self.game.display, 625, 370) )
		self.caselle.append( Casella(self.game.display, 525, 335) )
		self.caselle.append( Casella(self.game.display, 370, 340, "", 105, 95) )
		
		
	def draw_text(self, text, size, x, y):
		font = pygame.font.Font(self.game.font_name,size)
		text_surface = font.render(text, True, self.game.BLACK)
		text_rect = text_surface.get_rect()
		text_rect.center = (x,y)
		self.game.display.blit(text_surface,text_rect)

File: webApp/crazyGoose.js
const INFO_PL1 = "Tu (PL1)"
const INFO_COM = "Computer (COM)"
const INFO_DADO_PL1 = "Dado PL1: "
const INFO_DADO_COM = "Dado COM: "

function draw_text(ctx, text, size, color, x, y) {
    ctx.font = size.toString() + "px Arial";
    ctx.textAlign = "center"
    ctx.fillStyle = color
    ctx.fillText(text, x, y);
}

class Dado {
    tiraDado() {
        //num casuale tra 0 e 5 => +1 => num casuale tra 1 e 6
        return (Math.floor(Math.random() * 6) + 1)
    }
}

/*Non essendoci un widget/view Button neanche nel canvas ho dovuto "crearlo".
	Questo Button non è altro che un rettangolo
	con il testo centrato all'interno
*/
class Button {
    constructor(ctx, crazyGoose, game, x, y, width, height, text) {
        this.ctx = ctx
        this.crazyGoose = crazyGoose
        this.game = game
        this.x = x
        this.y = y
        this.width = width
        this.height = height

        this.ctx.beginPath()
        this.ctx.rect(x, y, width, height)
        this.ctx.stroke()
        draw_text(this.ctx, text, 14, "#000000", (x + this.width / 2), (y + this.height / 2 + 3))
    }

    controllaPosCursore(mouse) {
        //Controlla semplicemente che il cursore si trovi
        // all'interno del button.
        if (mouse[0] > this.x && mouse[0] < this.x + this.width &&
            mouse[1] > this.y && mouse[1] < this.y + this.height) {

            this.crazyGoose.tiraDado()
        }
    }

    detectMouseOver(mouse) {
        if (mouse[0] > this.x && mouse[0] < this.x + this.width &&
            mouse[1] > this.y && mouse[1] < this.y + this.height) {
            //codice per "evidenziare" il rettangolo
            // (per far capire all'utente che ci sta passando sopra)
            console.log("sopra");
        }
    }

}
class CrazyGoose {
    constructor(ctx, canvas, game) {
        this.ctx = ctx
        this.canvas = canvas
        this.game = game
        this.giocoPartito = false
    }

    start() {
        //Questo metodo è richiamato all'interno di un ciclo, quindi ho settato un flag,
        // per INIZIALIZZARE le variabili una volta soltanto.
        //(volendo il metodo può essere usato per RIAVVIARE il gioco
        //	N.B. BISOGNA RISETTARE A false il FLAG)
        if (this.giocoPartito == false) {
            this.giocoPartito = true
            this.partitaTerminata = false

            //Dentro questa lista ci saranno gli oggetti Casella
            this.caselle = []
                //Crea il percorso (casuale)
            this.percorso = new Percorso()

            this.valDadoCOM = "0"
            this.valDadoPL1 = "0"

            this.buttonDadoPL1 = null
            this.player = null
            this.com = null
            this.turnoCOM = null

            this.disegnaTutto()
        }
    }

    mousePremuto(mousePosition) {
        if (this.buttonDadoPL1 != null) {
            this.buttonDadoPL1.controllaPosCursore(mousePosition)
        }
    }

    mouseOver(mousePosition) {
        if (this.buttonDadoPL1 != null) {
            this.buttonDadoPL1.detectMouseOver(mousePosition)
        }
    }

    disegnaTutto() {
        console.log(this);
        this.ctx.clearRect(0, 0, this.canvas.width, this.canvas.height)
            /*this.ctx.beginPath()
            this.ctx.rect(0, 0, this.canvas.width, this.canvas.height)
            this.ctx.stroke()*/
        let flagPrimoGiro = (this.player == null && this.com == null)
            /*"ripulisce" tutto.
			  Purtroppo pygame funziona in questo modo, non posso "spostare" un elemento
			  posso solo ripartire da foglio bianco a disegnare*/
            //Riempe la lista di caselle, cioè "disegna" gli ellissi.
        this.posizionaLeCaselle()
            //Scrive nelle caselle il loro effetto
        this.riempiCaselle()

        draw_text(this.ctx, INFO_PL1, 15, "#000000", 50, 20)
        draw_text(this.ctx, INFO_DADO_PL1, 14, "#000000", 55, 60)

        //Il primo argomento è this xke necessita di questa classe per
        // richiamarne il metodo "tiraDado"
        this.buttonDadoPL1 = new Button(this.ctx, this, this.game, 100, 40, 35, 35, this.valDadoPL1)

        draw_text(this.ctx, INFO_COM, 14, "#000000", 930, 20)
        draw_text(this.ctx, (INFO_DADO_COM + this.valDadoCOM), 14, "#000000", 917, 60)

        if (this.player == null || this.player.posizione == 0) {
            //Se posizione == 0 vuol dire che è nella casella iniziale (quella "prima" del percorso)
            this.player = new Giocatore(this, this.ctx, this.caselle, this.percorso, "PL1")
        } else {
            this.player.spostaPedina(this.com.posizione)
        }

        if (this.com == null || this.com.posizione == 0) {
            //Se posizione == 0 vuol dire che è nella casella iniziale (quella "prima" del percorso)
            this.com = new Giocatore(this, this.ctx, this.caselle, this.percorso, "COM")
        } else {
            this.com.spostaPedina(this.player.posizione)
        }

        if (flagPrimoGiro == true) {
            // Decide chi incomincia, tira il dado e vede se il numero tirato è pari o dispari
            // (tra 1 e 6 ci sono 3 pari e 3 dispari, perciò 50% possbilità a testa)
            let random = (new Dado()).tiraDado()
            if (random % 2 == 0) {
                this.player.turnoMio = true
                this.com.turnoMio = false

                this.segnalaChiTocca(true)
            } else {
                this.player.turnoMio = false
                this.com.turnoMio = true

                //Fa giocare il COM. Aspetterà 2 sec e lancierà il dado....
                // il tutto in un altro thread (altro processo) xkè sennò
                // si bloccherebbe TUTTO per 2 secondi (a causa della sleep di 2 sec)

                this.toccaAlCOM()

                this.segnalaChiTocca(false)
            }
        }
    }

    tiraDado() {
        //Se la partita è terminata "blocca" la funzionalità del dado
        if (!this.partitaTerminata) {
            let numEstratto = (new Dado()).tiraDado()

            if (this.player.turnoMio) {
                this.avanzaPlayer1(numEstratto)

                //Per cambiare il numero uscito nel button nel html
                return numEstratto
            } else {
                this.toccaAlCOM()
            }
        } else { return 0 }
    }

    toccaAlCOM(numEstratto = -1) {
        //Se il num del dado non viene passato viene settato a -1 e allora lancia il dado
        if (numEstratto == -1) {
            numEstratto = (new Dado()).tiraDado()
        }

        //Non voglio bloccare il PL1 per 2 secondi, quindi se il COM ha un fermo non
        // faccio la sleep di 2 sec (dopo aver decrementato il fermo lancia il metodo
        // per far avanzare PL1)
        if (this.com.turniFermo == 0) {
            //dopo 2000ms (2sec) entra nella funzine, che a sua volta lancia la funzione avanzaCOM
            setTimeout(() => { this.avanzaCOM(numEstratto) }, 2000)
        } else {
            this.avanzaCOM(numEstratto)
        }
    }

    avanzaPlayer1(numEstratto) {
        if (this.player.turniFermo > 0) {
            // Se il PL1 ha beccato un fermo in precedenza deve "consumarlo"
            // (sarà come se avesse tirato l'avversario)
            this.player.turniFermo -= 1

            this.player.turnoMio = false
            this.com.turnoMio = true

            this.segnalaChiTocca(false)

            this.toccaAlCOM()
        } else {
            //(La prossima volta che verrà richiamato "disegnaTutto" il numero
            // del dado sarà cambiato)
            this.valDadoPL1 = numEstratto

            let toccaAncoraA_Me = this.player.avanza(numEstratto)

            if (!this.player.vincitore) {
                //avanza() ritorna this.turnoMio, perciò
                // se ritorna true non tocca all'avversario (gli setto false)
                // se ritorna false tocca all'avversario (gli setto true)
                this.com.turnoMio = !toccaAncoraA_Me

                //se prende un fermo ANNULLA il fermo dell'avversario 
                //(SENNÒ NESSUNO GIOCHEREBBE PIÙ per alcuni turni)
                if (this.player.turniFermo > 0) {
                    this.com.turniFermo = 0
                        //Ora il PL1 ha un fermo, quindi tocca all'avversario sicuro
                    this.segnalaChiTocca(false)

                    //Lancerà il dado, entrerà in avanzaPlayer1 che
                    // decrementerà il suo fermo e lancerà avanzaCOM
                    this.tiraDado()
                } else {
                    if (this.com.turniFermo > 0) {
                        //Se l'avversario ha un fermo al 100% tocca al PL1...
                        this.segnalaChiTocca(true)
                    } else {
                        //L'avversario non ha un fermo MA non è detto che tocchi a lui 
                        // (PL1 potrebbe aver preso un TIRA DI NUOVO) quindi controllo
                        if (this.com.turnoMio) {
                            this.segnalaChiTocca(false)

                            this.toccaAlCOM()
                        } else {
                            this.segnalaChiTocca(true)
                        }
                    }
                }
            } else {
                //Segnala PL1 come vincitore e fa TERMINARE la partita
                this.segnalaVincitore(true)
                this.partitaTerminata = true
            }
        }
    }

    avanzaCOM(numEstratto) {
        //(Per i commenti VEDI "avanzaPlayer1")

        if (this.com.turniFermo > 0) {
            this.com.turniFermo -= 1

            this.player.turnoMio = true
            this.com.turnoMio = false

            this.segnalaChiTocca(true)

            this.avanzaPlayer1(numEstratto)
        } else {
            this.valDadoCOM = numEstratto

            let toccaAncoraA_Me = this.com.avanza(numEstratto)

            this.player.turnoMio = !toccaAncoraA_Me
            if (!this.com.vincitore) {

                if (this.com.turniFermo > 0) {
                    this.player.turniFermo = 0
                    this.segnalaChiTocca(true)
                } else {
                    if (this.player.turniFermo > 0) {
                        this.segnalaChiTocca(false)
                        this.tiraDado()
                    } else {
                        if (this.player.turnoMio) {
                            this.segnalaChiTocca(true)
                        } else {
                            this.segnalaChiTocca(false)
                            this.toccaAlCOM()
                        }
                    }
                }
            } else {
                this.segnalaVincitore(false)
                this.partitaTerminata = true
            }
        }
    }

    segnalaChiTocca(turnoPlayer1) {
        let colorPL1 = ""
        let colorCOM = ""

        //Il colore è formato da R G B A dove la A è la trasparenza (0 "completamente trasparente" 255 "completamente visibile")
        if (turnoPlayer1) {
            colorPL1 = "#ff0000ff"
            colorCOM = "#ff000000"
        } else {
            colorPL1 = "#ff000000"
            colorCOM = "#ff0000ff"
        }

        this.ctx.beginPath()
        this.ctx.lineWidth = "2"
        this.ctx.strokeStyle = colorPL1
        this.ctx.rect(15, 5, 70, 20)
        this.ctx.stroke()

        this.ctx.beginPath()
        this.ctx.lineWidth = "2"
        this.ctx.strokeStyle = colorCOM
        this.ctx.rect(867, 5, 125, 20)
        this.ctx.stroke()

        //risetto al valore di default
        this.ctx.lineWidth = "1";
        this.ctx.strokeStyle = "#000000";
    }

    segnalaVincitore(haVintoPlayer1) {
        /*Codice per segnalare a chi tocca
        if(haVintoPlayer1):
        	this.lblInfoPl.configure(bg="orange")
        	this.lblInfoCom.configure(bg="grey")
        else:
        	this.lblInfoPl.configure(bg="grey")
        	this.lblInfoCom.configure(bg="orange")*/
    }

    riempiCaselle() {
        /*Cicla per ogni casella e controlla, se la posizione della casella
         c'è tra le chiavi del dizionario del percorso allora "carico" l'effetto
         (prendendo il codice contenuto nel valore con quella chiave, il codice della casella),
         altrimenti non metto nulla (uso il codice -1)*/
        //TODO controllare bene se gira per tutte le caselle
        for (let i = 1; i <= QTA_CASELLE_TOTALI; i++) {
            let casella = this.caselle[i - 1]
            try {
                casella.settaEffetto(this.percorso.dictCaselle[i])
            } catch (err) { //chiave inesistente (cioè posizione VUOTA)
                casella.settaEffetto(-1)
            }
        }
    }

    posizionaLeCaselle() {
        this.caselle.push(new Casella(this.ctx, 160, 590))
        this.caselle.push(new Casella(this.ctx, 270, 590))
        this.caselle.push(new Casella(this.ctx, 380, 590))
        this.caselle.push(new Casella(this.ctx, 490, 590))
        this.caselle.push(new Casella(this.ctx, 600, 590))
        this.caselle.push(new Casella(this.ctx, 710, 590))

        this.caselle.push(new Casella(this.ctx, 790, 535))
        this.caselle.push(new Casella(this.ctx, 840, 465))
        this.caselle.push(new Casella(this.ctx, 850, 395))
        this.caselle.push(new Casella(this.ctx, 840, 325))
        this.caselle.push(new Casella(this.ctx, 790, 255))

        this.caselle.push(new Casella(this.ctx, 710, 200))
        this.caselle.push(new Casella(this.ctx, 600, 170))
        this.caselle.push(new Casella(this.ctx, 490, 170))
        this.caselle.push(new Casella(this.ctx, 380, 170))
        this.caselle.push(new Casella(this.ctx, 270, 170))
        this.caselle.push(new Casella(this.ctx, 170, 210))

        this.caselle.push(new Casella(this.ctx, 125, 290))
        this.caselle.push(new Casella(this.ctx, 125, 375))
        this.caselle.push(new Casella(this.ctx, 125, 460))

        this.caselle.push(new Casella(this.ctx, 215, 520))
        this.caselle.push(new Casella(this.ctx, 325, 520))
        this.caselle.push(new Casella(this.ctx, 435, 520))
        this.caselle.push(new Casella(this.ctx, 545, 520))
        this.caselle.push(new Casella(this.ctx, 653, 505))

        this.caselle.push(new Casella(this.ctx, 735, 455))
        this.caselle.push(new Casella(this.ctx, 735, 370))
        this.caselle.push(new Casella(this.ctx, 680, 295))

        this.caselle.push(new Casella(this.ctx, 585, 250))
        this.caselle.push(new Casella(this.ctx, 475, 250))
        this.caselle.push(new Casella(this.ctx, 365, 250))

        this.caselle.push(new Casella(this.ctx, 265, 285))
        this.caselle.push(new Casella(this.ctx, 230, 360))
        this.caselle.push(new Casella(this.ctx, 270, 435))

        this.caselle.push(new Casella(this.ctx, 380, 450))
        this.caselle.push(new Casella(this.ctx, 490, 450))
        this.caselle.push(new Casella(this.ctx, 590, 430))

        this.caselle.push(new Casella(this.ctx, 625, 360))
        this.caselle.push(new Casella(this.ctx, 525, 325))
        this.caselle.push(new Casella(this.ctx, 375, 345, 65, 55))
    }

}

let canvas = document.getElementById("canvas")
var ctx = canvas.getContext("2d")

canvas.width = 1000
canvas.height = 650

let gioco = new CrazyGoose(ctx, canvas, null)
gioco.start()

//Aggiungo l'ascoltatore del click del mouse all'oggetto canvas (non potevo farlo nel file html xke mi serve l'oggetto gioco per lanciare un suo metodo)
canvas.addEventListener("mousedown", (event) => {

    //cliccato col tasto sx del mouse (se == 1 tasto centrale, se == 2 tasto dx)
    if (event.button == 0) {

        /*Con event.offsetX prendo le coord. X RISPETTO l'oggetto su cui c'è l'ascoltatore 
        (quindi non considera tutto quello spazio bianco, non occupato dal canvas)
        event.offsetY / event.clientY mi restituivano non so xke un valore TROPPO basso rispetto quello che mi aspettavo. Con event.pageY prendo le coord. Y RISPETTO la posizione nella PAGINA
        */
        let posMouse = new Array(event.offsetX, event.pageY)
        gioco.mousePremuto(posMouse)
    }
})